### PR TITLE
Fix bug when setting negative limits and using log scale

### PR DESCRIPTION
**Pull request by astrofrog**
This fixes an issue that caused the following error when setting negative limits on a log plot:

```
Traceback (most recent call last):
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/external/echo/qt/connect.py", line 480, in update_prop
    setattr(self._instance, self._prop, data_wrapper.data)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/external/echo/core.py", line 264, in __setattr__
    super(HasCallbackProperties, self).__setattr__(attribute, value)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/external/echo/selection.py", line 39, in __set__
    super(SelectionCallbackProperty, self).__set__(instance, value)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/external/echo/core.py", line 79, in __set__
    self.notify(instance, old, new)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/viewers/matplotlib/state.py", line 34, in notify
    super(DeferredDrawSelectionCallbackProperty, self).notify(*args, **kwargs)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/external/echo/core.py", line 128, in notify
    cback(new)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/core/state_objects.py", line 198, in _update_attribute
    self.update_values(attribute=self.component_id, use_default_modifiers=True)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/core/state_objects.py", line 378, in update_values
    self.set(lower=lower, upper=upper, percentile=percentile, log=log)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/core/state_objects.py", line 247, in set
    setattr(self, prop, value)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/external/echo/core.py", line 541, in __exit__
    p.notify(*args)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/viewers/matplotlib/state.py", line 23, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/external/echo/core.py", line 128, in notify
    cback(new)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File "/Users/tom/Dropbox/Code/Glue/glue/glue/viewers/matplotlib/viewer.py", line 187, in limits_to_mpl
    self.axes.set_xlim(x_min, x_max)
  File "/Users/tom/tmp/matplotlib/lib/matplotlib/axes/_base.py", line 3282, in set_xlim
    left = old_left
UnboundLocalError: local variable 'old_left' referenced before assignment
```

### CODE REVIEW

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -3240,10 +3240,11 @@ def set_xlim(self, left=None, right=None, emit=True, auto=False,
         left = self._validate_converted_limits(left, self.convert_xunits)
         right = self._validate_converted_limits(right, self.convert_xunits)
 
+        old_left, old_right = self.get_xlim()
+
         if left is None or right is None:
             # Axes init calls set_xlim(0, 1) before get_xlim() can be called,
             # so only grab the limits if we really need them.
```
Comment (anntzer) [inline lib/matplotlib/axes/_base.py]: Given the comment here, I'd assume that this actually needs to stay here (... for init order purposes) and you need to duplicate the call to get_xlim() under `if self.get_xscale("log"): ...`?

Comment (astrofrog) [reply lib/matplotlib/axes/_base.py]: Oops yes, I've fixed that

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -3249,7 +3249,10 @@ def set_xlim(self, left=None, right=None, emit=True, auto=False,
             if right is None:
                 right = old_right
 
-        if self.get_xscale() == 'log':
+        if self.get_xscale() == 'log' and (left <= 0 or right <= 0):
```
Comment (anntzer) [inline lib/matplotlib/axes/_base.py]: likely the same is needed for set_ylim?

**Comment by astrofrog** [comment]
I'm investigating the test failures

**Comment by astrofrog** [comment]
@anntzer - I think this is ready for a final review. The AppVeyor failure seems to be unrelated.

**Comment by anntzer** [comment]
Milestoning as 3.2 as I think the bug is only on master?

**Comment by astrofrog** [comment]
@anntzer - I didn't run into the issue with older versions so I think 3.2 makes sense.

**Pull request merged**